mindmeld/auto_annotator.py

```diff
@@ -835,5 +895,282 @@ def valid_entity_check(self, entity):
+    """Custom Annotator class used to generate annotations."""
+
+    def __init__(self, app_path, config=None, language=None, locale=None):
+        """Initializes an annotator.
+
+        Args:
+            app_path (str): The location of the MindMeld app
+            config (dict, optional): A config object to use. This will
+                override the config specified by the app's config.py file.
+        """
+        super().__init__(app_path=app_path, config=config)
+        self.language = language or self.language
+        self.locale = locale or self.locale
+
+        self.en_annotator = SpacyAnnotator(
+            app_path=self.app_path, config=self.config, language=ENGLISH_LANGUAGE_CODE
+        )
+        if self.language != ENGLISH_LANGUAGE_CODE:
+            self.duckling_annotator = self._get_duckling_annotator()
+            self.non_en_annotator = SpacyAnnotator(
+                app_path=self.app_path,
+                config=self.config,
+                language=self.language,
+                locale=self.locale,
+            )
+
+    def _get_duckling_annotator(self):
+        if "translator" in self.config:
+            return TranslationDucklingAnnotator(
+                self.app_path, self.config, self.en_annotator
+            )
+        return NoTranslationDucklingAnnotator(self.app_path, self.config)
+
+    def parse(self, sentence, entity_types=None, language=None, locale=None, **kwargs):
+        """
+        Args:
+            sentence (str): Sentence to detect entities.
+            entity_types (list): List of entity types to parse. If None, all
+                    possible entity types will be parsed.
+            language (str, optional): Language as specified using a 639-1/2 code.
+            locale (str, optional): The locale representing the ISO 639-1 language code and \
+                ISO3166 alpha 2 country code separated by an underscore character.
+        Returns: entities (list): List of entity dictionaries.
+        """
+        language = language or self.language
+        locale = locale or self.locale
+
+        if self.language == ENGLISH_LANGUAGE_CODE:
+            return self.en_annotator.parse(sentence, entity_types=entity_types)
+        non_en_spacy_entities = self.non_en_annotator.parse(
+            sentence, entity_types=entity_types, language=language, locale=locale
+        )
+        duckling_entities = self.duckling_annotator.parse(
+            sentence, entity_types=entity_types, language=language, locale=locale
+        )
+        merged_entities = MultiLingualAnnotator._resolve_conflicts_entity_dicts(
+            non_en_spacy_entities, duckling_entities
+        )
+        return merged_entities
+
+    @property
+    def supported_entity_types(self):  # pylint: disable=W0236
+        """
+        Returns:
+            supported_entity_types (list): List of supported entity types.
+        """
+        return SPACY_ANNOTATOR_SUPPORTED_ENTITIES
```

Comment:
This probably needs to also be conditioned on language, no? Especially since this annotator depends on duckling candidates.

Reply:
Yes, that's correct. Planning to return the union of the entities supported by Spacy and Duckling. There isn't any documentation that includes a list of system entities supported by language for MindMeld's version of Duckling. The closest data available is in "allDimensions" in Duckling/Dimensions files. For example, Duckling/Dimensions/BG.hs:

```
  | ( allDimensions
  | ) where
  |  
  | import Duckling.Dimensions.Types
  |  
  | allDimensions :: [Some Dimension]
  | allDimensions =
  | [ This Distance
  | , This Duration
  | , This Numeral
  | , This Ordinal
  | ]
```

Would need some way to collect this data and then map dimensions to a list of entities. Open to suggestions for tackling this

Reply:
Added a dictionary that maps language codes to the supported entities. Added code to determine the supported entities for each class, this list is dynamically created based on the initialized language.